source/server/listener_manager_impl.cc

```diff
@@ -567,6 +567,15 @@ void ListenerManagerImpl::addListenerToWorker(Worker& worker, ListenerImpl& list
       }
       if (success) {
         stats_.listener_create_success_.inc();
+
+        if (listener.reusePort()) {
+          // The shared socket created for acquiring port number when port is configured as 0
+          // can be closed here.
```

Comment:
Worker threads run concurrently, main thread posts per-worker listener creation to all workers, the order is not guaranteed.

Reply:
Envoy::Server::WorkerImpl::addListener() is called on main thread, so socket can be created in it and posted to worker thread. In this way, Envoy::Network::ConnectionHandler::addListener() should be split into two:

- Envoy::Network::ConnectionHandler::addTcpListener()
- Envoy::Network::ConnectionHandler::addUdpListener()

And ActiveUdpListener and ActiveTcpListener's ctor don't call `onfig.listenSocketFactory().getListenSocket()` to create socket any more.

Sound good?

Reply:
One point just hit me. Actually even with non-zero port, unless we create all the sockets first and then post to different workers, there is always unstable hashing for connections opened between when the corresponding socket starts listening and when another socket is created. So I think we need to create all the sockets according to concurrency in ListenerManager first before we start listening on any of them if we are reusing port.

Reply:
For TCP I don't see any issue with the unstable hashing you mentioned, after a connection accepted, a random port  number will be allocated for establishing the connection.
For UDP, packet are stateless datagram, does it matter which socket a packet lands on?

The only issue with ephemeral socket I can see is with UDP, after bind() called(happens during NetworkSocket creation), packets may be delivered to the socket, then after the ephemeral socket closed, the packets looks like dropped on data path, packet being dropped is acceptable for UDP. Even all reuse port UDP sockets got created on main thread and posted to worker threads, before worker threads actually start receiving packets, if the traffic is heavy, some packets may be dropped. Please correct me if I'm wrong.

Reply:
TCP with a temporary socket might also has some issue. What if a new connection is dispatched to the temporary socket before it's closed? We don't have a listener for that socket.

Reply:
> What if a new connection is dispatched to the temporary socket before it's closed? 

the temporary socket doesn't call listen(), no connection will be dispatched to it.

Reply:
> For UDP, packets are dispatched to one of the reuse_port sockets based on the 4 tuple. The mapping is between the 4 tuple and the index of the socket in a socket map for the reused port. If there are change in that map, the index of each socket might be changed (depending on kernel implementation). So the packet with the same 4 tuple will be mapped to different listening socket.

Even all reuse port UDP sockets are created on main thread at same time, these sockets are bound one by one, during this procedure, the problem you described could happen as well, right? I feel this is becoming a listener readiness issue, that's, before a listener's reuse port sockets are all bound(that's, hash map is stable), no UDP traffic should land on Envoy. 

Looks like QUIC has solution for this issue; for UDP proxying, considering UDP server/client must tolerate packet-dropping, packet-reordering, I guess this doesn't cause real impact in most of cases.

Reply:
> > What if a new connection is dispatched to the temporary socket before it's closed?
> 
> the temporary socket doesn't call listen(), no connection will be dispatched to it.

TBH, I don't know the kernel behavior for TCP where we have N sockets bound to a port via SO_REUSEPORT and (N - 1) of them are listening. Better to get some kernel person to confirm...

Reply:
`man listen` has a description about its behavior:

> listen() marks the socket referred to by sockfd as a passive socket,
> that is, as a socket that will be used to accept incoming connection
> requests using accept(2).

see http://man7.org/linux/man-pages/man2/listen.2.html

Reply:
> Even all reuse port UDP sockets are created on main thread at same time, these sockets are bound one by one, during this procedure, the problem you described could happen as well, right? I feel this is becoming a listener readiness issue, that's, before a listener's reuse port sockets are all bound(that's, hash map is stable), no UDP traffic should land on Envoy.
> Looks like QUIC has solution for this issue; for UDP proxying, considering UDP server/client must tolerate packet-dropping, packet-reordering, I guess this doesn't cause real impact in most of cases.

This might not be a packet loss or reordering issue from proxying Envoy's perspective. Packets from the same client might be proxied by two different worker threads to two different backend services. Because the two listeners on different threads will regard them as two individual traffic flow.

It is a listener readiness issue indeed. If we can guarantee that all the listeners start receiving traffic at the same time after socket binding and cleaning up, yes, opening and closing temporary shared socket in between doesn't have real impact. @mattklein123, @alyssawilk is this guaranteed?

The same problem can happen during listener shutdown too. Because any socket close can cause the socket mapping change. We need to guarantee that sockets are closed at the same time. I feel that UDP SO_REUSEPORT might need special handling at this point.

Current approach seems to work for TCP in my eyes. So if you just want to focus on TCP case, feel free to leave UDP socket as it.

Reply:
The temporary socket created here will be handed over to first worker thread which calls ListenSocketFactory::getListenSocket() now, that's, no ephemeral port closing anymore.

Reply:
This sounds safer! Please also document the design somewhere and the rationale behind, including TCP and UDP (if you want to get your hands dirty) initial starts up and tear down, as well as config update. If even it's not supported in current implementation, it's better to mention it at least. I think it's also better to get kernel team looped in for the design.

Reply:
We need to maintain port 0 ephemeral binding for both integration testing and some use cases, e.g. an Envoy binds port 0, then does service registration with its discovered port.